v0.12 testing: TypeSet Count Mismatch

I'm running into an issue with the OpenStack Provider where the length of a TypeSet is being calculated incorrectly.

You can see the test output here:

http://logs.openlabtesting.org/logs/59/559/68760b7eefc5536645c911b5af7319596166c604/check/terraform-provider-openstack-acceptance-test/febc61e/

via the job-output.txt.gz file.

The error is:

```
--- FAIL: TestAccNetworkingV2Port_noFixedIP (38.33s)
    testing.go:568: Step 2 error: Check failed: Check 2/2 error: openstack_networking_port_v2.port_1: Attribute 'all_fixed_ips.#' expected "0", got "1"
```

However, if I add a debug line above [here](https://github.com/terraform-providers/terraform-provider-openstack/blob/master/openstack/resource_openstack_networking_port_v2.go#L342):

```go
log.Printf("FOO: expandNetworkingPortFixedIPToStringSlice(p.FixedIPs))
d.Set("all_fixed_ips", expandNetworkingPortFixedIPToStringSlice(p.FixedIPs))
```

The result is a correct `[]string{}`, but the following test is failing:

```
resource.TestCheckResourceAttr("openstack_networking_port_v2.port_1", "all_fixed_ips.#", "0"),
```

What's interesting is that it is Step 2 of the test that fails. Step 0 runs this same fixture and passes:

Step 0: Port with No Fixed IPs
Step 1: Update the port to have 1 Fixed IP
Step 2: Update the port to have 0 Fixed IPs
- Failure: `all_fixed_ips.#` expected 0, got 1

The API is really returning no Fixed IPs and changing the test to look like this allows the test to pass: https://github.com/terraform-providers/terraform-provider-openstack/pull/559/commits/9316c2e78a60fae59f60ab7c1e377bb9d5083463

Please let me know if you need any more information or if there's anything I can try.

--- a/builtin/providers/test/resource_list.go
+++ b/builtin/providers/test/resource_list.go
@@ -58,21 +58,54 @@ func testResourceList() *schema.Resource {
 					},
 				},
 			},
+			"dependent_list": {
+				Type:     schema.TypeList,
+				Optional: true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"val": {
+							Type:     schema.TypeString,
+							Required: true,
+						},
+					},
+				},
+			},
+			"computed_list": {
+				Type:     schema.TypeList,
+				Computed: true,
+				Elem:     &schema.Schema{Type: schema.TypeString},
+			},
 		},
 	}
 }
 
 func testResourceListCreate(d *schema.ResourceData, meta interface{}) error {
 	d.SetId("testId")
-	return nil
+	return testResourceListRead(d, meta)
 }
 
 func testResourceListRead(d *schema.ResourceData, meta interface{}) error {
+	fixedIps := d.Get("dependent_list")
+
+	// all_fixed_ips should be set as computed with a CustomizeDiff func, but
+	// we're trying to emulate legacy provider behavior, and updating a
+	// computed field was a common case.
+	ips := []interface{}{}
+	if fixedIps != nil {
+		for _, v := range fixedIps.([]interface{}) {
+			m := v.(map[string]interface{})
+			ips = append(ips, m["val"])
+		}
+	}
+	if err := d.Set("computed_list", ips); err != nil {
+		return err
+	}
+
 	return nil
 }
 
 func testResourceListUpdate(d *schema.ResourceData, meta interface{}) error {
-	return nil
+	return testResourceListRead(d, meta)
 }
 
 func testResourceListDelete(d *schema.ResourceData, meta interface{}) error {

--- a/builtin/providers/test/resource_list_test.go
+++ b/builtin/providers/test/resource_list_test.go
@@ -276,3 +276,45 @@ resource "test_resource_list" "foo" {
 		},
 	})
 }
+
+func TestResourceList_addRemove(t *testing.T) {
+	resource.UnitTest(t, resource.TestCase{
+		Providers:    testAccProviders,
+		CheckDestroy: testAccCheckResourceDestroy,
+		Steps: []resource.TestStep{
+			resource.TestStep{
+				Config: strings.TrimSpace(`
+resource "test_resource_list" "foo" {
+}
+				`),
+				Check: resource.ComposeTestCheckFunc(
+					resource.TestCheckResourceAttr("test_resource_list.foo", "computed_list.#", "0"),
+					resource.TestCheckResourceAttr("test_resource_list.foo", "dependent_list.#", "0"),
+				),
+			},
+			resource.TestStep{
+				Config: strings.TrimSpace(`
+resource "test_resource_list" "foo" {
+	dependent_list {
+		val = "a"
+	}
+}
+				`),
+				Check: resource.ComposeTestCheckFunc(
+					resource.TestCheckResourceAttr("test_resource_list.foo", "computed_list.#", "1"),
+					resource.TestCheckResourceAttr("test_resource_list.foo", "dependent_list.#", "1"),
+				),
+			},
+			resource.TestStep{
+				Config: strings.TrimSpace(`
+resource "test_resource_list" "foo" {
+}
+				`),
+				Check: resource.ComposeTestCheckFunc(
+					resource.TestCheckResourceAttr("test_resource_list.foo", "computed_list.#", "0"),
+					resource.TestCheckResourceAttr("test_resource_list.foo", "dependent_list.#", "0"),
+				),
+			},
+		},
+	})
+}

--- a/helper/plugin/grpc_provider.go
+++ b/helper/plugin/grpc_provider.go
@@ -431,7 +431,7 @@ func (s *GRPCProviderServer) ReadResource(_ context.Context, req *proto.ReadReso
 		// here we use the prior state to check for unknown/zero containers values
 		// when normalizing the flatmap.
 		stateAttrs := hcl2shim.FlatmapValueFromHCL2(stateVal)
-		newInstanceState.Attributes = normalizeFlatmapContainers(stateAttrs, newInstanceState.Attributes, false)
+		newInstanceState.Attributes = normalizeFlatmapContainers(stateAttrs, newInstanceState.Attributes)
 	}
 
 	if newInstanceState == nil || newInstanceState.ID == "" {
@@ -572,7 +572,7 @@ func (s *GRPCProviderServer) PlanResourceChange(_ context.Context, req *proto.Pl
 	// now we need to apply the diff to the prior state, so get the planned state
 	plannedAttrs, err := diff.Apply(priorState.Attributes, block)
 
-	plannedAttrs = normalizeFlatmapContainers(priorState.Attributes, plannedAttrs, false)
+	plannedAttrs = normalizeFlatmapContainers(priorState.Attributes, plannedAttrs)
 
 	plannedStateVal, err := hcl2shim.HCL2ValueFromFlatmap(plannedAttrs, block.ImpliedType())
 	if err != nil {
@@ -808,7 +808,7 @@ func (s *GRPCProviderServer) ApplyResourceChange(_ context.Context, req *proto.A
 	// here we use the planned state to check for unknown/zero containers values
 	// when normalizing the flatmap.
 	plannedState := hcl2shim.FlatmapValueFromHCL2(plannedStateVal)
-	newInstanceState.Attributes = normalizeFlatmapContainers(plannedState, newInstanceState.Attributes, true)
+	newInstanceState.Attributes = normalizeFlatmapContainers(plannedState, newInstanceState.Attributes)
 
 	// We keep the null val if we destroyed the resource, otherwise build the
 	// entire object, even if the new state was nil.
@@ -819,6 +819,7 @@ func (s *GRPCProviderServer) ApplyResourceChange(_ context.Context, req *proto.A
 	}
 
 	newStateVal = normalizeNullValues(newStateVal, plannedStateVal, false)
+
 	newStateVal = copyTimeoutValues(newStateVal, plannedStateVal)
 
 	newStateMP, err := msgpack.Marshal(newStateVal, block.ImpliedType())
@@ -995,7 +996,7 @@ func pathToAttributePath(path cty.Path) *proto.AttributePath {
 // allows a provider to set an empty computed container in the state without
 // creating perpetual diff. This can differ slightly between plan and apply, so
 // the apply flag is passed when called from ApplyResourceChange.
-func normalizeFlatmapContainers(prior map[string]string, attrs map[string]string, apply bool) map[string]string {
+func normalizeFlatmapContainers(prior map[string]string, attrs map[string]string) map[string]string {
 	isCount := regexp.MustCompile(`.\.[%#]$`).MatchString
 
 	// While we can't determine if the value was actually computed here, we will
@@ -1008,11 +1009,6 @@ func normalizeFlatmapContainers(prior map[string]string, attrs map[string]string
 	for k, v := range prior {
 		if isCount(k) && (v == "0" || v == hcl2shim.UnknownVariableValue) {
 			zeros[k] = true
-		}
-
-		// fixup any 1->0 conversions that happened during Apply
-		if apply && isCount(k) && v == "1" && attrs[k] == "0" {
-			attrs[k] = "1"
 		}
 	}
 
@@ -1291,7 +1287,7 @@ func normalizeNullValues(dst, src cty.Value, plan bool) cty.Value {
 		// If the dst is nil, and the src is known, then we lost an empty value
 		// so take the original.
 		if dst.IsNull() {
-			if src.IsWhollyKnown() && !plan {
+			if src.IsWhollyKnown() && src.LengthInt() == 0 && !plan {
 				return src
 			}
 			return dst

--- a/helper/plugin/grpc_provider_test.go
+++ b/helper/plugin/grpc_provider_test.go
@@ -716,7 +716,7 @@ func TestNormalizeFlatmapContainers(t *testing.T) {
 		},
 	} {
 		t.Run(strconv.Itoa(i), func(t *testing.T) {
-			got := normalizeFlatmapContainers(tc.prior, tc.attrs, false)
+			got := normalizeFlatmapContainers(tc.prior, tc.attrs)
 			if !reflect.DeepEqual(tc.expect, got) {
 				t.Fatalf("expected:\n%#v\ngot:\n%#v\n", tc.expect, got)
 			}

